File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        self.res = root.val
        def dfs(node):
            if not node:
                return 0
            left = max(dfs(node.left), 0)
            right = max(dfs(node.right), 0)
            # update res if splitting is greater
            self.res = max(left + right + node.val, self.res)

            # return max of left, right or node
            return max(left + node.val, right + node.val, node.val)

        dfs(root)
        return self.res
                

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {TreeNode}
     */
    lowestCommonAncestor(root, p, q) {
        console.log(root)
        if ((p.val <= root.val && root.val <= q.val) || (q.val <= root.val && root.val <= p.val)) {
            return root
        }
        if (root.left && p.val <= root.val && q.val <= root.val) {
            return this.lowestCommonAncestor(root.left, p, q)
        }
        if (root.right && p.val >= root.val && q.val >= root.val) {
            return this.lowestCommonAncestor(root.right, p, q)
        }
        
    }
}
